flink-core/src/main/java/org/apache/flink/api/common/accumulators/AbstractAccumulatorRegistry.java

```
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.common.accumulators;
+
+import java.io.Serializable;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Accumulator registry encapsulates user-defined accumulators.
+ */
+public abstract class AbstractAccumulatorRegistry {
+	/* User-defined accumulator values stored for the executing task. */
+	protected final Map<String, Accumulator<?, ?>> userAccumulators =
+		new ConcurrentHashMap<>(4);
+
+	/**
+	 * Adds an ordinary accumulator.
+	 */
+	public <V, A extends Serializable> void addAccumulator(String name, Accumulator<V, A> accumulator) {
+		if (userAccumulators.containsKey(name)) {
+			throw new UnsupportedOperationException("The accumulator '" + name
+				+ "' already exists and cannot be added.");
+		}
+
+		userAccumulators.put(name, accumulator);
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **39 and 44**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 If the key already exists, the code should throw an error at line: **40**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.